client/lib/trigram/index.js

```diff
@@ -0,0 +1,134 @@
+/**
+ * External dependencies
+ */
+import LRU from 'lru';
+
+/**
+ * Convert a string into a list of trigrams. A trigram is a 3 character substring.
+ * We use a special token for the beginning and the end of the string.
+ *
+ * Example: trigrams("Hello") = [ "_BEGIN_He", "Hel", "ell", "llo", "lo_END_" ]
+ *
+ * @param {string} str A string to convert to trigrams.
+ * @returns {Array} A List of trigrams.
+ */
+const trigrams = ( str ) => {
+	const n = 3;
+	const grams = [];
+	const input = [ '_BEGIN_', ...str.split( '' ), '_END_' ];
+	for ( let i = 0; i <= input.length - n; i++ ) {
+		grams.push( input.slice( i, i + n ).join( '' ) );
+	}
+	return grams;
+};
+
+/**
+ * Convert a list of trigrams into a lookup table.
+ * The keys are trigrams, the values are counts.
+ *
+ * Example: "mississippi"
+ * grams_to_lookup([ "_BEGIN_mi", "mis", "iss", "ssi", "sis", "iss", "ssi", "sip", "ipp", "ppi", "pi_END_" ]) =
+ *     { "_BEGIN_mi": 1, "mis": 1, "iss": 2, "ssi": 2, "sis": 1, "sip": 1, "ipp": 1, "ppi": 1, "pi_END_": 1 }
+ *
+ * @param {string} gram_list A list of trigrams, like [ "_BEGIN_He", "Hel", "ell", "llo", "lo_END_" ]
+ * @returns {object} A lookup table of trigram frequency.
+ */
+const grams_to_lookup = ( gram_list ) => {
+	const lookup = {};
+	for ( const gram of gram_list ) {
+		if ( gram in lookup ) {
+			lookup[ gram ] += 1;
+		} else {
+			lookup[ gram ] = 1;
+		}
+	}
+	return lookup;
+};
+
+/**
+ * Convert a trigram lookup table into a real number magnitude, (aka l2 norm or
+ * euclidean distance).  Think of each trigram lookup table as a vector, where
+ * there are ~26 * 26 * 26 dimensions ("aaa" ... "zzz")
+ *
+ * Example:
+ * lookup_to_magnitude(grams_to_lookup('hi')) = 1.414215... (sqrt of 2)
+ *
+ * @param {object} lookup A lookup table of trigram frequency.
+ * @returns {number} Magnitude of the vector
+ */
+const lookup_to_magnitude = ( lookup ) => {
```

Comment:
I'd love to see some tests for those functions.

Reply:
Done :+1: